031-statefulset-removal.md

```diff
@@ -0,0 +1,252 @@
+Without the StatefulSet limitations, we should have a lot more flexibility to mix the node types and their roles.
+
+### Stretch clusters
+
+Pods managed by a single StatefulSet always exist only on single Kubernetes cluster.
+If you want to have a Kafka cluster which would stretch multiple Kubernetes clusters, you would need to create a StatefulSet on each of them and have it create the Pods on each given cluster.
+But you would not be able easily move the brokers between the Kubernetes clusters since they would depend on the StatefulSet index numbers to generate their broker IDs.
+By introducing _StrimziPodSets_ (for more details and explanation, check the _Proposal_ section below), you can have different StrimziPodSets in each Kubernetes cluster and just delete a pod on one Kubernetes cluster and move it to another cluster by modifying the corresponding StrimziPodSets.
+
+## Proposal
+
+The StatefulSets should be replaced with a new Strimzi custom resource called `StrimziPodSet`.
+StrimziPodSet will be used only internally.
+Users will not be expected to interact with.
+
+The StrimziPodSet custom resource and its structure might be subject to change during the _alpha_ phase.
+But once it reaches the _beta_ phase, it will be treated as any other Strimzi custom resource and provide backwards compatibility and versioning in case any backwards incompatible changes would be required.
+See the _Feature Gate_ section below for more details about the feature gate phases and the expected timelines.
+
+The StrimziPodSets will have a following structure:
+* The usual metadata, kind and API identifications
+* A `.spec` section with:
+    * Pod selector to be able to find the pods managed by this StrimziPodSet
+    * List of Pods (full Pod YAML definition including metadata section, but without status section)
+* A status section with:
+    * Number of managed, current and ready Pods
+    * List of error conditions in case of problems
+    * Observed generation
+
+The StrimziPodSet will be used very similarly to the StatefulSets:
+* The Pods will be always created in the same namespace as the pod set
+* Garbage collection will be used to delete the Pods
+
+Instead of using a Pod templates, and generating Pods with fixed names, StrimziPodSet will use full Pod structures including custom names.
+The configuration retains the usefulness of a StatefulSet, but removes the limitations we have identified.
+
+The StrimziPodSet will be created by the `KafkaAssemblyOperator` class in the same place in the reconciliation as the StatefulSet.
+This is also where the StrimziPodSet will be patched when configuration changes.
+
+A new controller will be developed to reconcile the StrimziPodSet resources.
+It will run as part of the Strimzi cluster operator container.
+It will follow a more lightweight controller approach rather than the long-running _heavy_ approach of our current assembly operators.
```

Comment:
what does this exactly mean?

Reply:
That is supposed to mean what is described in the sentences below this one:
* It will be designed to iterate quickly rather than wait
* It will not wait for Pod readiness but instead get a new event when the Pod becomes ready
* It will consume events about the StrimziPodSets and their Pods and react to them
* It will run in its own thread and use a work queue to queue the events and the reconciliations to avoid working on too many reconciliations in parallel

I changed the text a bit to make it more clear.